### PR TITLE
docs: Add coding guidelines

### PR DESCRIPTION
**Pull request by diehbria**
## Overview
Add coding guidelines and other minor documentation fixes.

Made the branding more consistently 'IoT App Kit' (As opposed to IoT Application Kit)

Revision 1: Fixed some minor formatting issues and added more examples in the guidelines
Revision 2: Addressed comments
## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: docs/CodingGuidelines.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+1. Tests names should be descriptive and from a users perspective:  
+   *GOOD NAME EXAMPLE*: “it creates new user when new user form is submitted”  
+   *BAD NAME EXAMPLE*: “button dispatches onClick event”
+
+1. Tests should, at minimum, cover all core feature use-cases and identifiable edge-cases or boundary conditions.
+
+1. Code must be written in a consistent style, utilizing the shared eslint and stylelint configurations https://github.com/awslabs/iot-app-kit/blob/main/.eslintrc.js https://github.com/awslabs/iot-app-kit/blob/main/.stylelintrc
+1. Pull requests to feature branches must contain only a single, well formatted commit that adheres to https://www.conventionalcommits.org/en/v1.0.0/
+1. Pull requests to main must have correct change logs which properly document migrations, feature changes, and most importantly breaking change.
+1. Pull requests to main must either be a single commit, or be a series of commits with prior code review plus a single commit for any necessary changes to rebase on top of main
```
Comment (sydopy) [inline docs/CodingGuidelines.md:41]: would it be possible to put an example of what this would look like?

**Review by sydopy — DISMISSED** [review]

File: docs/CodingGuidelines.md (hunk 1 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:41]: in the next iteration i will add a lot more detail on the git management portion and go into more detail on how to manage the commit history

File: docs/CodingGuidelines.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+    The being said, there’s nothing wrong with local UI state managed directly in a component. For example, for a collapsable menu, it will probably make sense to just store the open state directly in the component. Do not blindly pull all state into a centralized state manager as a reducer.
+
+    *EXAMPLE OF BAD STATE MANAGEMENT*:
+
+    UI components storing state of multiple async calls, coordinating loading status, and pagination
+
+## IoT App Kit component requirements
+
+1. Components will utilize https://cloudscape.design/ for primitive components, unless a good case is made not to. The default is to use cloudscape.
```
Comment (boweihan) [inline docs/CodingGuidelines.md:110]: nit: Cloudscape should be capitalized / linked to

File: docs/CodingGuidelines.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+1. Styling should adhere to https://cloudscape.design/ design principles and guidelines.
+1. Web components will be exported within https://github.com/awslabs/iot-app-kit/tree/main/packages/components, with a iot- prefix. i.e. iot-line-chart. Their styles should be contained within `"@iot-app-kit/components/styles.css".
+1. React components will be exported within https://github.com/awslabs/iot-app-kit/tree/main/packages/react-components with no prefix, i.e. LineChart
+1. Components can also be created in other repositories, but must ultimately be exported from either @iot-app-kit/components and @iot-app-kit/react-components (or both)
```
Comment (boweihan) [inline docs/CodingGuidelines.md:114]: looks like number is off here again, shouldn't all be 1?

File: docs/CodingGuidelines.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+1. Components should support various data sources, via either a query or queries property.
+   for example,
+     ```
+      <iot-line-chart  
+      queries={[query.timeSeriesData(...)]}  
+      ...  
+    />  
+    ```
+1. It is ok to have data-source specific features. Not all data-source features will be generic in nature, but that shouldn’t prevent us from exposing them to make it easy for customers to utilize the feature
```
Comment (boweihan) [inline docs/CodingGuidelines.md:154]: nit: missing period

In some of the bullet points below there are commas instead of periods. Numbering looks off here too.

File: docs/CodingGuidelines.md (hunk 3 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:114]: i'm having a fight with my IDE on this, will fix it

File: docs/CodingGuidelines.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+   Compare writing to other written examples within https://docs.aws.amazon.com/iot-sitewise/ as a primary source on how to write the documentation, as well as cross-referencing other documentation found within https://github.com/awslabs/iot-app-kit/tree/main/docs.
+
+1. Ensure that all documentation is correctly referenced in the table of contents, which can be found at https://github.com/awslabs/iot-app-kit/blob/main/README.md.
+
+   In-depth guidelines on documentation: https://alpha-docs-aws.amazon.com/awsstyleguide/latest/styleguide/awsstyleguide.pdf#aws-implementation-guides. This can be a useful reference, but you are not expected to read through and be 100% adherent to these guidelines.
+1. Components must have high quality testing, including integration tests which utilize the component in the way a user does. An example of a good example integration test can be found in the [@iot-app-kit/source-iotsitewise package](https://github.com/awslabs/iot-app-kit/blob/main/packages/source-iotsitewise/src/time-series-data/data-source.spec.ts#L562-L603)
```
Comment (tjuranek) [inline docs/CodingGuidelines.md:14]: Thoughts on combining the first two items? Both a documentation related.

File: docs/CodingGuidelines.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+1. Tests should be written to maximize the chance that the test would actually catch a real issue, i.e. the likelihood of preventing a regression. This is the primary value of a test.  Useful talk on testing: https://www.youtube.com/watch?v=Fha2bVoC8SE3.
+
+2. Tests should also be written to minimize the pain they inflict on the development cycle. Every test represents some amount of burden, through the necessity to run it, maintain it, and alter the test as the codebase morphs, however some tests inflict more pain than others.  
+   *EXAMPLES OF PAIN*:
+  - tests failing when nothing broke
+  - flaky tests
+  - snapshot tests that vary on machine
+  - tests that require refactoring when irrelevant portions of the codebase are altered
+
+1. Tests should be written to not fail when things are not broken (this helps reduce inflicted pain). This can be achieved through techniques such as not over asserting.
```
Comment (tjuranek) [inline docs/CodingGuidelines.md:25]: Duplicate of the first "example of pain". Any way to combine these or change the formatting a bit?

File: docs/CodingGuidelines.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    <iot-line-chart size=“SM” ... /> 
+    ```
+
+    BAD EXAMPLE*:
+    ```  
+    <iot-line-chart size={Size.SM} ... /> 
+    ```  
+
+14. Make the code consistent with the existing code base
+15. Avoid mutating stuff. Prefer immutability
```
Comment (tjuranek) [inline docs/CodingGuidelines.md:67]: nit: "avoid mutations when possible. prefer immutability"

File: docs/CodingGuidelines.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
+
```
Comment (tjuranek) [inline docs/CodingGuidelines.md:1]: General comment on whole doc:

I think the separation of topics could be better. If I want to review the guidelines for testing, I can't go to a "testing" section. I have to kinda skim through the whole doc for it. Same with pull requests.

Could we group the items by topic? "Testing", "Pull Requests", "Syntax/Linting", "Coding Best Practices", etc.

**Review by tjuranek — APPROVED** [review]
Mostly LGTM! Any of my comments could be made in a later pull request. Great start!

File: docs/CodingGuidelines.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+# Coding guidelines
+
+Guidelines that code authors and code reviewers are expected to adhere to in the development of IoT App Kit.
+
+## General guidelines
+
+1. Documentation must be written in a style that is consistent with the rest of the documentation (based on a AWS documentation style)  
+   Compare writing to other written examples within https://docs.aws.amazon.com/iot-sitewise/ as a primary source on how to write the documentation, as well as cross-referencing other documentation found within https://github.com/awslabs/iot-app-kit/tree/main/docs.
+
+1. Ensure that all documentation is correctly referenced in the table of contents, which can be found at https://github.com/awslabs/iot-app-kit/blob/main/README.md.
```
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md:11]: that's the normal way to write github documentation as the viewer automatically increments it. Though i'm a little inconsistent since my IDE is trying to automatically number things sometimes.

File: docs/CodingGuidelines.md (hunk 5 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:14]: I'm trying to avoid making long guidelines with many 'rules' as than things get lost. but I agree w/ your idea to add some additional organization so things are easier to find and there could be a 'documentation' section

File: docs/CodingGuidelines.md (hunk 4 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:154]: utilizing best practice for numbering here: https://google.github.io/styleguide/docguide/style.html#use-lazy-numbering-for-long-lists

File: docs/CodingGuidelines.md [hunk 10]
```diff
@@ -0,0 +1,174 @@
+
+# Coding guidelines
+
+Guidelines that code authors and code reviewers are expected to adhere to in the development of IoT App Kit.
+
+## General guidelines
+
+1. Documentation must be written in a style that is consistent with the rest of the documentation (based on a AWS documentation style)
```
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+   In-depth guidelines on documentation: https://alpha-docs-aws.amazon.com/awsstyleguide/latest/styleguide/awsstyleguide.pdf#aws-implementation-guides. This can be a useful reference, but you are not expected to read through and be 100% adherent to these guidelines.
+1. Components must have high quality testing, including integration tests which utilize the component in the way a user does. An example of a good example integration test can be found in the [@iot-app-kit/source-iotsitewise package](https://github.com/awslabs/iot-app-kit/blob/main/packages/source-iotsitewise/src/time-series-data/data-source.spec.ts#L562-L603)
+
+1. Tests should be written to maximize the chance that the test would actually catch a real issue, i.e. the likelihood of preventing a regression. This is the primary value of a test.  Useful talk on testing: https://www.youtube.com/watch?v=Fha2bVoC8SE3.
+
+2. Tests should also be written to minimize the pain they inflict on the development cycle. Every test represents some amount of burden, through the necessity to run it, maintain it, and alter the test as the codebase morphs, however some tests inflict more pain than others.
```
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+18. Consider performance where it matters
+
+    In some scenarios, improving the performance will add complexity, while not actually effecting the end performance seen by the users. In this case, always opt for simplicity.
+
+19. Embrace refactoring
+
+    Refactoring is crucial to the health of any code base. Fearlessly refactor where it can help improve simplicity. Often times code when originally written misses the mark. Do not be afraid to refactor code written by other authors.
+20. State is the root of complexity
+
+    Minimize the amount of state is required, and avoid entangling state management through the entire codebase.  Often times state should be pushed towards the boundary (such as in a reducer).
```
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md (hunk 2 above)
Comment (diehbria) [reply docs/CodingGuidelines.md]: Fixed

File: docs/CodingGuidelines.md (hunk 4 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:154]: fixed missing period

File: docs/CodingGuidelines.md (hunk 7 above)
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md (hunk 8 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:1]: Yes i think that is a good idea, i'd like to have that done in a future revision

File: docs/CodingGuidelines.md (hunk 6 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:25]: see https://github.com/awslabs/iot-app-kit/pull/208#discussion_r945917917

File: docs/CodingGuidelines.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+7. Pull requests to main from a feature branch should maintain the git history as contained in the feature branch
+8. Pull requests to main must have the correct versioning updates, following https://semver.org/
+9. Pull request descriptions should include a reasonably detailed summary of changes and highlight any architectural decisions introduced.
+10. Reviewers are expected to enforce a high standard on:
+    1. Consistent code to rest of codebase
+    2. High quality tests that are not painful to maintain
+    3. Architectural and big picture aspects of how the change fits in
+11. It is ok to build out work incremental within a feature branch, but not in main.
+
+    For example, it is OK to have a PR merged into a feature branch with missing test cases and add those in the next PR. This helps allow people to get code in more frequently rather than having long lived local branched
```
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md:50]: going to need to do a major revision of the git section in a future revision with more explicit directions

File: docs/CodingGuidelines.md [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+5. Pull requests to feature branches must contain only a single, well formatted commit that adheres to https://www.conventionalcommits.org/en/v1.0.0/
+6. Pull requests to main must have correct change logs which properly document migrations, feature changes, and most importantly breaking change. [example commit](https://github.com/awslabs/iot-app-kit/commit/9e09d6229dc5d27309766db4717d5a250e91bdd7)
+7. Pull requests to main from a feature branch should maintain the git history as contained in the feature branch
+8. Pull requests to main must have the correct versioning updates, following https://semver.org/
+9. Pull request descriptions should include a reasonably detailed summary of changes and highlight any architectural decisions introduced.
+10. Reviewers are expected to enforce a high standard on:
+    1. Consistent code to rest of codebase
+    2. High quality tests that are not painful to maintain
+    3. Architectural and big picture aspects of how the change fits in
+11. It is ok to build out work incremental within a feature branch, but not in main.
```
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+    In some scenarios, improving the performance will add complexity, while not actually effecting the end performance seen by the users. In this case, always opt for simplicity.
+
+19. Embrace refactoring
+
+    Refactoring is crucial to the health of any code base. Fearlessly refactor where it can help improve simplicity. Often times code when originally written misses the mark. Do not be afraid to refactor code written by other authors.
+20. State is the root of complexity
+
+    Minimize the amount of state is required, and avoid entangling state management through the entire codebase.  Often times state should be pushed towards the boundary (such as in a reducer).
+
+    The being said, there’s nothing wrong with local UI state managed directly in a component. For example, for a collapsable menu, it will probably make sense to just store the open state directly in the component. Do not blindly pull all state into a centralized state manager as a reducer.
```
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md (hunk 3 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:114]: https://google.github.io/styleguide/docguide/style.html#use-lazy-numbering-for-long-lists following best practice on this, need to utilize it more consistently

**Review by tjuranek — APPROVED** [review]

File: docs/CodingGuidelines.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (boweihan) [reply docs/CodingGuidelines.md:11]: TIL

File: docs/CodingGuidelines.md [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+2. Tests should be written to minimize the pain they inflict on the development cycle. Every test represents some amount of burden, through the necessity to run it, maintain it, and alter the test as the codebase morphs, however some tests inflict more pain than others.  
+   *EXAMPLES OF PAIN*:
+  - tests failing when nothing broke
+  - flaky tests
+  - snapshot tests that vary on machine
+  - tests that require refactoring when irrelevant portions of the codebase are altered
+
+1. Tests should be written to not fail when things are not broken (this helps reduce inflicted pain). This can be achieved through techniques such as not over asserting.  
+   *GOOD ASSERTION EXAMPLE*:
```
Comment (boweihan) [inline docs/CodingGuidelines.md:26]: nit: line break above this

File: docs/CodingGuidelines.md [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+
+   In-depth guidelines on documentation: https://alpha-docs-aws.amazon.com/awsstyleguide/latest/styleguide/awsstyleguide.pdf#aws-implementation-guides. This can be a useful reference, but you are not expected to read through and be 100% adherent to these guidelines.
+1. Components must have high quality testing, including integration tests which utilize the component in the way a user does. An example of a good example integration test can be found in the [@iot-app-kit/source-iotsitewise package](https://github.com/awslabs/iot-app-kit/blob/main/packages/source-iotsitewise/src/time-series-data/data-source.spec.ts#L562-L603)
+
+1. Tests should be written to maximize the chance that the test would actually catch a real issue, i.e. the likelihood of preventing a regression. This is the primary value of a test.  Useful talk on testing: https://www.youtube.com/watch?v=Fha2bVoC8SE3.
+
+2. Tests should be written to minimize the pain they inflict on the development cycle. Every test represents some amount of burden, through the necessity to run it, maintain it, and alter the test as the codebase morphs, however some tests inflict more pain than others.  
+   *EXAMPLES OF PAIN*:
```
Comment (boweihan) [inline docs/CodingGuidelines.md:19]: nit: line break above this

File: docs/CodingGuidelines.md [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+
+1. Tests should be written to not fail when things are not broken (this helps reduce inflicted pain). This can be achieved through techniques such as not over asserting.  
+   *GOOD ASSERTION EXAMPLE*:  
+   expect(result).toEqual(expect.objectContaining({ error: null })
+
+   *BAD ASSERTION EXAMPLE*:  
+   every test asserting on every property over and over whether or not they are relevant to the purpose of the test
+
+2. Tests names should be descriptive and from a users' perspective:  
+   *GOOD NAME EXAMPLE*: “it creates new user when new user form is submitted”
```
Comment (boweihan) [inline docs/CodingGuidelines.md:33]: nit: line breaks

File: docs/CodingGuidelines.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (boweihan) [reply docs/CodingGuidelines.md:50]: yeah fair, this is a typo though - "local branched" -> "local branches"

File: docs/CodingGuidelines.md [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+1. Tests should be written to minimize the pain they inflict on the development cycle. Every test represents some amount of burden, through the necessity to run it, maintain it, and alter the test as the codebase morphs, however some tests inflict more pain than others.
+
+   *EXAMPLES OF PAIN*:
+    - tests failing when nothing broke
+    - flaky tests
+    - snapshot tests that vary on machine
+    - tests that require refactoring when irrelevant portions of the codebase are altered
+
+1. Tests should be written to not fail when things are not broken (this helps reduce inflicted pain). This can be achieved through techniques such as not over asserting.  
+   *GOOD ASSERTION EXAMPLE*:
```
Comment (boweihan) [inline docs/CodingGuidelines.md:27]: Line break above this

File: docs/CodingGuidelines.md [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+   *GOOD ASSERTION EXAMPLE*:  
+   
+   `expect(result).toEqual(expect.objectContaining({ error: null })`
+
+   *BAD ASSERTION EXAMPLE*:  
+   
+   every test asserting on every property over and over whether or not they are relevant to the purpose of the test
+
+1. Tests names should be descriptive and from a users' perspective:  
+1.
```
Comment (boweihan) [inline docs/CodingGuidelines.md:36]: There shouldn't be a number here

File: docs/CodingGuidelines.md [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+    ```
+
+    *BAD EXAMPLE*:
+    ```  
+    <iot-line-chart size={Size.SM} ... /> 
+    ```  
+
+14. Make the code consistent with the existing code base
+15. Avoid mutations when possible. Prefer immutability
+    *GOOD EXAMPLE*:
```
Comment (boweihan) [inline docs/CodingGuidelines.md:73]: Need a line break here, it's inlining on the previous line

File: docs/CodingGuidelines.md [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+
+    *EXAMPLE OF BAD STATE MANAGEMENT*:
+
+    UI components storing state of multiple async calls, coordinating loading status, and pagination
+
+## IoT App Kit component requirements
+
+1. Components will utilize [Cloudscape](https://cloudscape.design/) for primitive components, unless a good case is made not to. The default is to use Cloudscape.
+1. Styling should adhere to https://cloudscape.design/ design principles and guidelines.
+1. Web components will be exported within https://github.com/awslabs/iot-app-kit/tree/main/packages/components, with a iot- prefix. i.e. iot-line-chart. Their styles should be contained within `"@iot-app-kit/components/styles.css".
```
Comment (boweihan) [inline docs/CodingGuidelines.md:120]: You're missing a final backtick at the end of this line, so it doesn't come out as a code block

File: docs/CodingGuidelines.md [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+14. Make the code consistent with the existing code base
+15. Avoid mutations when possible. Prefer immutability
+    *GOOD EXAMPLE*:  
+    Returns updated object without mutating the input
+     ```  
+     (myObject) => ({ ...myObject, name: trimName(myObject.name (http://myobject.name/)) }) 
+    ```  
+    *BAD EXAMPLE*:  
+    Mutates input to method (DO NOT DO THIS)  
+    `(myObject) => { myObject.name (http://myobject.name/) = trimName(myObject.name) } `
```
Comment (boweihan) [inline docs/CodingGuidelines.md:80]: Use triple quotes here for the code block for consistency with the good example

**Review by boweihan — CHANGES_REQUESTED** [review]

File: docs/CodingGuidelines.md (hunk 21 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:77]: fixed

File: docs/CodingGuidelines.md (hunk 20 above)
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md (hunk 19 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:29]: fixed

File: docs/CodingGuidelines.md (hunk 16 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:29]: fixed

File: docs/CodingGuidelines.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (diehbria) [reply docs/CodingGuidelines.md]: fixed

File: docs/CodingGuidelines.md (hunk 18 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:39]: fixed

File: docs/CodingGuidelines.md (hunk 17 above)
Comment (diehbria) [reply docs/CodingGuidelines.md:20]: fixed

**Review by boweihan — APPROVED** [review]
